algosdk/atomic_transaction_composer.py

```diff
@@ -165,67 +188,156 @@ def add_method_call(
-                        "cannot add a transaction with nonzero group ID"
-                    )
                 txn_list.append(method_args[i])
-            elif len(app_args) > self.MAX_ABI_APP_ARG_LIMIT:
-                # Pack the remaining values as a tuple
-                additional_types.append(arg.type)
-                additional_args.append(method_args[i])
             else:
-                encoded_arg = arg.type.encode(method_args[i])
-                app_args.append(encoded_arg)
+                if abi.is_abi_reference_type(arg.type):
+                    current_type = abi.UintType(8)
+                    if arg.type == abi.ABIReferenceType.ACCOUNT:
+                        account_arg = str(method_args[i])
+                        if account_arg == sender:
+                            current_arg = 0
+                        elif account_arg in accounts:
+                            current_arg = accounts.index(account_arg) + 1
+                        else:
+                            current_arg = len(accounts) + 1
+                            accounts.append(account_arg)
+                    elif arg.type == abi.ABIReferenceType.ASSET:
+                        asset_arg = int(method_args[i])
+                        if asset_arg in foreign_assets:
+                            current_arg = foreign_assets.index(asset_arg)
+                        else:
+                            current_arg = len(foreign_assets)
+                            foreign_assets.append(asset_arg)
+                    elif arg.type == abi.ABIReferenceType.APPLICATION:
+                        app_arg = int(method_args[i])
+                        if app_arg == app_id:
+                            current_arg = 0
+                        elif app_arg in foreign_apps:
+                            current_arg = foreign_apps.index(app_arg) + 1
+                        else:
+                            current_arg = len(foreign_apps) + 1
+                            foreign_apps.append(app_arg)
+                    else:
+                        # Shouldn't reach this line unless someone accidentally
+                        # adds another foreign array arg
+                        raise error.AtomicTransactionComposerError(
+                            "cannot recognize {} as a foreign array arg".format(
+                                arg.type
+                            )
+                        )
+                else:
+                    current_type = arg.type
+                    current_arg = method_args[i]
 
-        if additional_args:
-            remainder_args = abi.TupleType(additional_types).encode(
-                additional_args
+                raw_types.append(current_type)
+                raw_values.append(current_arg)
+
+        if len(accounts) > self.FOREIGN_ACCOUNT_LIMIT:
+            raise error.AtomicTransactionComposerError(
+                "foreign accounts exceed app limit {}".format(
```

Comment:
```suggestion
                "foreign accounts exceed account number limit {}".format(
```

Reply:
I will delete this check entirely because I think it is slightly better to have this check on a lower level (e.g. `algod`) as discussed